core/conn/unixodbc/odbc/odbcclient/unixcli/cli/ctosqlconv.cpp

```diff
@@ -157,852 +157,982 @@ unsigned long ODBC::Ascii_To_Interval_Helper(char *source,
+                    usTmp = (USHORT)dTmp;
+                    DataPtr = &usTmp;
+                    DataLen = sizeof(USHORT);
+                    break;
+                case SQLTYPECODE_SMALLINT:
+                    sTmp = (SHORT)dTmp;
+                    DataPtr = &sTmp;
+                    DataLen = sizeof(SHORT);
+                    break;
+
+                case SQLTYPECODE_INTEGER_UNSIGNED:
+                    ulTmp = (ULONG_P)dTmp;
+                    DataPtr = &ulTmp;
+                    DataLen = sizeof(ULONG_P);
+                    break;
+                case SQLTYPECODE_INTEGER:
+                    lTmp = (LONG)dTmp;
+                    DataPtr = &lTmp;
+                    DataLen = sizeof(LONG);
+                    break;
+
+                case SQLTYPECODE_LARGEINT_UNSIGNED:
+                    tempVal64 = (unsigned __int64)dTmp;
+                    DataPtr = &tempVal64;
+                    DataLen = sizeof(unsigned __int64);
+                    break;
+                case SQLTYPECODE_LARGEINT:
+                    tempVal64 = (__int64)dTmp;
+                    DataPtr = &tempVal64;
+                    DataLen = sizeof(__int64);
+                    break;
+
+                default:
+                    return IDS_07_006;
+            }
+        }
+        else
+        {
+            if( targetUnsigned && negative )
+                return IDS_22_003_02;   //negValue in unsigned column
+
+            if (targetScale)
+            {
+                for (i = 0,tempVal64 = 1; i < targetScale ;  i++)
+                    tempVal64 *= 10;
+                tempVal64 = tempVal64 * integralPart;
+                decimalDigits = 0;
+                if (decimalPart > 0)
+                    decimalDigits = getDigitCount(decimalPart);
+                scaleOffset = 0;
+                if (leadZeros < targetScale)
+                    scaleOffset = targetScale - decimalDigits - leadZeros;
+                if (scaleOffset < 0)
+                {
+                    //NUMERIC_VALUE_OUT_OF_RANGE_ERROR
+                    return IDS_22_003;
+                }
+
+                for (i =0, tempScaleVal64 = decimalPart ; i < scaleOffset ; i++)
+                    tempScaleVal64 *= 10;
+                tempVal64 += tempScaleVal64;
+            }
+            else
+            {
+                //NUMERIC_DATA_TRUNCATED_ERROR
+                if (decimalPart != 0)
+                    retCode = IDS_01_S07;
+                tempVal64 = integralPart;
+            }
+            if (negative)
+                tempVal64 = -tempVal64;
+
+            switch( SQLDataType )
             {
+                case SQLTYPECODE_TINYINT_UNSIGNED:
+                    if (tempVal64 < 0)
+                        return IDS_22_003_02;
+                    if ((UCHAR)tempVal64 > UCHAR_MAX)
+                        return IDS_22_003;
+                    utTmp = (UCHAR)tempVal64;
+                    if  (tempVal64 != utTmp)
```

Comment:
This test makes the bug at 564 almost unobservable. If tempVal64 were UCHAR_MAX+1, instead of error IDS_22_003, it looks like we will catch the problem here and get error IDS_01_S07 instead.

Reply:
IDS_22_003 means numeric out of range while IDS_01_S07 means fractional truncation.